Iframe Service

jagthedrummer opened the pull request:
This is a freshly rebased branch of https://github.com/Vestorly/torii/pull/206

The basic overview, which is described in an update to the README is:

## Using an iframe instead of a popup

You can configure torii to use an in-page iframe instead of a separate
popup window for authentication. This can be done on either a global or
a per-provider basis.

To change this globally set the `remoteServiceName` variable in the main
torii config to be `'iframe'`.

```JavaScript
/* jshint node: true */
// config/environment.js
module.exports = function(environment) {
  var ENV = {
    /* ... */
    torii: {
      remoteServiceName: 'iframe',
      providers: { /* ... */ }
    }
  };
  return ENV;
};
```

If you only want the iframe for a single provider you can include the
`remoteServiceName` value in the configuration for that provider.

```JavaScript
/* jshint node: true */
// config/environment.js
module.exports = function(environment) {
  var ENV = {
    /* ... */
    torii: {
      // a 'session' property will be injected on routes and controllers
      sessionServiceName: 'session',
      providers: {
        'mycorp-oauth2': {
          remoteServiceName: 'iframe'
          /* ... */
        }
      }
    }
  };
  return ENV;
};
```

Once your provider has been configured you need to tell torii where to
apend the iframe when you call `session.open`.

For instance:

```JavaScript
this.get("session")
  .open("mycorp-oauth2",{ iframeParent : '#signin-modal-content'})
  .then(function(){
    // ...
  });
```

Of course, the element with the id of `signin-modal-content` needs to be
currently in the DOM.

/end of README excerpt

---

Instead of requiring the `iframeParent` argument to be passed to `open` we might include a component that people can include in a template to tell torii where to put the iframe. Something like:

```handlebars
{{torii-iframe-placeholder}}
```

I took a stab at making this happen, but so far I haven't been able to get the component to be useable in the host app. I suspect that this is related to the custom build system and the fact that I don't entirely understand how it works and how it's different from ember-cli based add-ons. Would love some pointers on that note.

jagthedrummer commented:
@mixonic, I just realized that Friday managed to sneak past without me ever thinking about trying to sync up with you on this. Maybe sometime this week?

jagthedrummer commented:
Just pinging this again. :smiley:

mixonic commented on lib/torii/bootstrap/torii.js:
```diff
@@ -30,9 +32,15 @@ export default function(application) {
 
-  application.inject('torii-provider', 'popup', 'torii-service:popup');
+  var remoteServiceName = configuration.remoteServiceName || 'popup';
+  if(remoteServiceName === 'iframe'){
```
Can we do this lazily instead? Add a `popup` property on the base provider that looks up the proper thing based on the `config`?

mixonic replied:
I think it would improve/remove the need for stuff in `walk-providers` as well. Instead of setting the popup, it should just inject a non-instantiated property with the name of the provider (or even with the config)

jagthedrummer replied:
@mixonic If I'm following what you're saying here, would that mean that we want to do something like this here?

```javascript
var remoteServiceName = configuration.remoteServiceName || 'popup';
application.inject('torii-provider', 'remoteServiceProviderName',
  'torii-service:' + remoteServiceName);
```

And then we'd have a computed property in `lib/torii/providers/base.js` that somehow uses the value of the `'remoteServiceProvider'` property to look up either the popup or iframe service? If so, how do I look up a service that's registered on the application from inside a provider?

```javascript
  popup: computed('remoteServiceProviderName', function(){
    return ?.lookup(this.get('remoteServiceProviderName'));
  })
```

mixonic replied:
Right, something along those lines. you can `inject` with `instantiate: false` if you want to just inject the remote service name directly as a prop.

mixonic commented on lib/torii/mixins/ui-service-mixin.js:
```diff
@@ -0,0 +1,138 @@
+import UUIDGenerator from 'torii/lib/uuid-generator';
+import PopupIdSerializer from 'torii/lib/popup-id-serializer';
+import ParseQueryString from 'torii/lib/parse-query-string';
+//import {stringifyOptions, prepareOptions, parseMessage} from 'torii/helpers/services-helper';
```
Can drop this if un-used

mixonic commented on lib/torii/mixins/ui-service-mixin.js:
```diff
@@ -0,0 +1,138 @@
+
+var on = Ember.on;
+
+
```
Some whitespace here :-)

mixonic commented on lib/torii/mixins/ui-service-mixin.js:
```diff
@@ -0,0 +1,138 @@
+
+function parseMessage(url, keys){
+  var parser = ParseQueryString.create({url: url, keys: keys}),
+      data = parser.parse();
```
Please split these declarations into multiple statements, with commas they are tricky to drop a breakpoint on.

mixonic commented on lib/torii/mixins/ui-service-mixin.js:
```diff
@@ -0,0 +1,138 @@
+    this._super.apply(this, arguments);
+    options = options || {};
+    this.remoteIdGenerator = options.remoteIdGenerator || UUIDGenerator;
+    console.log('this.remoteIdGenerator = ', this.remoteIdGenerator);
```
`console.log`

mixonic commented on lib/torii/mixins/ui-service-mixin.js:
```diff
@@ -0,0 +1,138 @@
+
+  init: function(options){
+    console.log('creating with options : ', options);
+    // may not need these next two lines : https://github.com/Vestorly/torii/commit/f5531d325b95df825ae763ef064c5acc36be8858#diff-8b7b6e15c618f576b36e529bef3c661bL51
```
`_super` would be needed, the `options` line can be dropped if you rewrite the logic for `options.remoteIdGenerator`

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,29 @@
+
+  // Open an iframe window.
+  openRemote: function(url, pendingRequestKey, options){
+    this.remote = Ember.$('<iframe src="'+url+'" id="torii-iframe"></iframe>'); //window.open(url, pendingRequestKey, optionsString);
```
Comment unneeded

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,29 @@
+  openRemote: function(url, pendingRequestKey, options){
+    this.remote = Ember.$('<iframe src="'+url+'" id="torii-iframe"></iframe>'); //window.open(url, pendingRequestKey, optionsString);
+    var iframeParent = 'body';
+    if(options && options.iframeParent){
```
Space it out :-) `if (options && options.iframeParent) {`

Regardless, I think we should assert the option is passed instead of defaulting to `body`. It seems like a fairly useless default, just good for demos.

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,29 @@
+
+var Iframe = Ember.Object.extend(Ember.Evented, UiServiceMixin, {
+
+  // Open an iframe window.
```
`Append an iframe`

mixonic commented on test/tests/unit/services/popup-test.js:
```diff
@@ -50,7 +50,7 @@ asyncTest("open resolves based on popup window", function(){
   var mockWindow = null
 
-  popup = Popup.create({popupIdGenerator: buildPopupIdGenerator(popupId)});
+  popup = new Popup({remoteIdGenerator: buildPopupIdGenerator(popupId)});
```
We should definitely use `create`. Invocation of Ember objects with `new` is not actually public API (I thought it was so if you see any left here, my fault :-p)

bantic commented on README.md:
````diff
@@ -318,6 +318,70 @@ via the `torii` property injected onto _routes_, or the `session` property
+```
+
+Once your provider has been configured you need to tell torii where to
+apend the iframe when you call `session.open`.
````
"append"

mixonic commented:
@jagthedrummer THANKS. This is looking really good man. Sorry it has taken to long to think it through and find the time.

I'd like to suggest one more largish adjustment. Right now, before an iframe is opened, the developer must ensure an element with the correct id (or selector) is on the page. This leaks some domish stuff where it doesn't feel appropriate.

Instead, let's add a component `{{torii-iframe}}` that can only be used once (should assert if used more than that) and can manage the DOM parts of the iframe.

jagthedrummer commented:
@mixonic @bantic In this branch/PR I've addressed all of the issues that you guys had pointed out with the exception of the "lazy registration" issue. I've worked on that and have something that is different, and I think a little more like what you're asking for. Unfortunately is doesn't actually work, and I'm not entirely sure why.

This commit is what I have going so far: https://github.com/jagthedrummer/torii/commit/64de9e1c532e5b88f03667789c9853c52efe140e

The issue is that even though the value of `'remoteServiceName'` is being set to `'iframe'` in `walk-providers`, when I actually call `get('popup')` on the provider, inside that computed property the value of `'remoteServiceName'` always comes back as `'popup'`.

One other issue I've discovered with this approach is that for the stuff in `walk-providers` to actually work for a custom provider you have to create an initializer for your provider that registers it with the application so that torii can find it. Registering the provider with the application isn't mentioned anywhere in the docs for creating a customer provider, so that would mean that any existing custom providers would have to be updated before they could be used with the iframe service. The "non-lazy" current code in this PR (not in the commit linked above) should be transparently backward compatible with any existing providers.

I feel like there's probably some way to do the "lazy" thing in a way that 1) actually works and 2) is transparently backward compatible with existing providers. What am I missing?

mixonic commented:
@jagthedrummer I'll attempt an implementation tomorrow and PR against your branch :-)

mixonic commented:
with lazy popup service discovery in https://github.com/jagthedrummer/torii/pull/1

jagthedrummer commented:
@mixonic Thanks for the PR on the lazy lookup. I just merged it into this branch/PR and everything is working perfectly in my test app. I think this should be all ready to go. Thanks for the help!

mixonic commented on README.md:
```diff
@@ -318,6 +318,101 @@ via the `torii` property injected onto _routes_, or the `session` property
+  route.controller.set('signingIn',true);
+  // We need to user Ember.run.next to make sure that the placeholder
+  // component has been added to the DOM before session.open is called
+  Ember.run.next(this,function(){
```
`Ember.run.schedule('afterRender', this, function() {`, not `next`.

mixonic commented on lib/torii/redirect-handler.js:
```diff
@@ -21,7 +22,10 @@ var RedirectHandler = Ember.Object.extend({
 
-        windowObject.close();
+        var remoteServiceName = configuration.remoteServiceName || 'popup';
+        if(remoteServiceName === 'popup'){
```
this seems bad. If the default config is `popup` but the specific running remote service is `iframe` the `close` would still be called.

jagthedrummer replied:
Ah, yeah. Hmm... Seems like here in `redirect-handler` we need to somehow discover which provider we're handling and be able to find the config set for just that provider. Do you have a suggestion on how to approach that?

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,31 @@
+
+var on = Ember.on;
+
+var Iframe = Ember.Object.extend(Ember.Evented, UiServiceMixin, {
```
Should extend `Ember.Service`

jagthedrummer replied:
This was based on the existing code for `services/popup.js` which does not extend `Ember.Service`. Should both of them be updated to extend that?

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,31 @@
+var Iframe = Ember.Object.extend(Ember.Evented, UiServiceMixin, {
+
+  // For some reason having this init in the UiServiceMixin will not work...
+  init: function(){
```
should call `this._super(...arguments);`

jagthedrummer replied:
Does `services/popup.js` need this also? https://github.com/Vestorly/torii/blob/master/lib/torii/services/popup.js#L51-L53

mixonic commented on lib/torii/services/iframe.js:
```diff
@@ -0,0 +1,31 @@
+
+var Iframe = Ember.Object.extend(Ember.Evented, UiServiceMixin, {
+
+  // For some reason having this init in the UiServiceMixin will not work...
```
This comment can maybe go away, or be improved? What is it supposed to tell a future traveler?

jagthedrummer replied:
Yeah, it could be better. It's basically supposed to mean "even though this identical code shows up in both popup.js and iframe.js DON'T try to move it to the UiServiceMixin that they both inherit from".

mixonic commented on lib/torii/services/popup.js:
```diff
@@ -40,120 +37,31 @@ function prepareOptions(options){
-}
-
-var Popup = Ember.Object.extend(Ember.Evented, {
+var Popup = Ember.Object.extend(Ember.Evented, UiServiceMixin, {
```
Hm, can this also be a service? That they are objects is just old code from before services came in.
